Feature/variables revisited

protich opened the pull request:
Add `%team` and `%staff` to base variables
Add "object" level variables for staff, team and dept.

Do NOT merge yet: see issue #222 for pending discussion.

protich committed:
Add %staff and %team to base variables.
Change %assigned to mean assignees (staff or/and team)

protich committed:
Add "object" level variables to staff, team and dept.
Example use case %staff.name  %team.name ... etc.

protich committed:
Merge upstream develop

protich committed:
Add variables replacer class - used to resolve and replace variables within email templates

protich committed:
Support asVar - returns an equivalent of the name of the object

protich committed:
Merge filters revisited branch

protich committed:
Add 'asVar' function - returns staff's name

protich committed:
Add array based variables assignment
Allow empty variable - or whatever objects return

protich committed:
Merge branch 'issue/270' into feature/variables-revisited

protich committed:
Add support for multidimentional array in assign & replaceVars routines

protich committed:
Add "asVar" support to help topic object

protich committed:
Use the new variable replacer

protich committed:
copy ticket transfer alert template

protich committed:
Add support for new variables to email templates + upgrade patch

protich commented:
Doing a pull request for review purposes - I'll rebase it to chance the schema signature once attachment PR is merged.

greezybacon commented on include/class.variable.php:
```diff
@@ -0,0 +1,145 @@
+    function _parse($text) {
+
+        $input = $text;
+        if(!preg_match_all('/'.$this->start_delim.'([A-Za-z\._]+)'.$this->end_delim.'/', $input, $result))
```
You don't have to escape the `.` inside brackets, and PHP allows number chars in the function names as well. I would suggest `[A-Za-z_][\w._]+`

greezybacon commented on include/class.ticket.php:
```diff
@@ -1084,51 +1087,87 @@ function onOverdue($whine=true, $comments="") {
+                break;
+                
+            case 'clientlink':
+            case 'client_link':
```
I really think we should be clear and simple with this. There should be only one way to do something. It's either `clientlink` or `client_link`, `auth_code` or `auth_token`, etc. I think here that flexibility  will just make things confusing. Personally, I like the all lower-case values with underscores

greezybacon commented on include/class.ticket.php:
```diff
@@ -1216,8 +1252,8 @@ function transfer($deptId, $comments, $alert = true) {
                 if(!is_object($staff) || !$staff->isAvailable() || in_array($staff->getEmail(),$sentlist)) continue;
-                $alert = str_replace("%staff",$staff->getFirstName(),$body);
-                $email->send($staff->getEmail(),$subj,$alert);
+                $alert = str_replace('%{recipient}',$staff->getFirstName(), $msg['body']);
```
Why not something like

``` php
$alert = $this->replaceVars($msg['body'], array('recipient' => $staff));
```

and use `%{recipient.first_name}` in the body as the token?

protich replied:
The plan (my private plan??) is to create a mailer class  ... e.g

``` php
$mailer = new osTicketMailer($email,....);
$mailer->send($recipients, $msg, $options);
```

The mailer will create a single instance of VariableReplacer and use it for every recipient.  In addition it will help abstract all the for loops used to fire alerts to staff.

Please note that recipient can be just an email address, staff object or client object - so probably not good idea to do `%{recipient.first_name}` in all templates. 

**PS**: Mailer class actually already exists - it's `class.mailer.php`. Subconscious mind at work!

greezybacon commented on include/class.ticket.php:
```diff
@@ -1527,12 +1558,8 @@ function postNote($title,$note,$alert=true,$poster='') {
-            $body=$this->replaceTemplateVars($msg['body']);
-            $subj=$this->replaceTemplateVars($msg['subj']);
-            $body = str_replace('%note',"$title\n\n$note",$body);
-            # TODO: Support a variable replacement of the staff writing the
```
Why did you drop the TODO?

protich replied:
Not exactly sure why - but I *think" my reasoning was - a note can be posted by system as part of an event log - adding `%{staff.name}` to internal note template will be problematic in such contexts.

greezybacon commented on include/class.variable.php:
```diff
@@ -0,0 +1,145 @@
+<?php
+/*********************************************************************
+    class.filter.php
```
Whoops

protich replied:
I copied the copyright header - and forgot to change the name. Good catch.

greezybacon commented on include/upgrader/sql/15b30765-7f35de42.patch.sql:
```diff
@@ -0,0 +1,409 @@
+/*
+ * @version=1.6RC2+
```
1.7RC2+

greezybacon commented on include/class.ticket.php:
```diff
@@ -1527,12 +1558,8 @@ function postNote($title,$note,$alert=true,$poster='') {
-            # TODO: Support a variable replacement of the staff writing the
-            #       note
+                   
+            $msg = $this->replaceVars($msg, array('note' => "$title\n\n$note"));
```
I vote for both `%{title}` and `%{note}` replacements, and just change the `%note` currently in the database to `%{title}\n\n%{note}`.

protich replied:
I agree with you - but _feel_ like it should be `%{note.title}` & `%{note.message}`. Which will give us the flexibility of doing `%{note.staff.name}` or `%{note.poster}` - resolving context issues when compared to `%{staff.name}` or even `%{poster}` for note poster. See my other comment on the same issue.

That said, it will open a pandora box on the need to add classes for note, message & response objects (which should extend something like class.threaditem.php). You get the idea... I simply chickened out.

I'll implement it _your way_ for now, and then we'll revisit variables post RC3 release. Please? Pretty please??

greezybacon commented on include/ajax.content.php:
```diff
@@ -56,7 +56,9 @@ function ticket_variables() {
                     <tr><td>%status</td><td>Status</td></tr>
```
Why didn't you add the braces here?

protich replied:
It's planned - made the pull request for the initial review.

greezybacon commented on include/class.dept.php:
```diff
@@ -277,6 +281,23 @@ function delete() {
     }
 
+    //Replace staff's variables
+    function replaceVariables($text, $prefix='dept') {
```
This function doesn't appear to be used

greezybacon commented on include/class.staff.php:
```diff
@@ -502,6 +506,26 @@ function delete() {
     }
 
+    //Replace staff's variables
+    function replaceVariables($text, $prefix='staff') {
```
This function doesn't appear to be used

greezybacon commented on include/class.team.php:
```diff
@@ -169,13 +173,28 @@ function delete() {
     }
 
+    //Replace staff's variables
+    function replaceVariables($text, $prefix='team') {
```
This function doesn't appear to be used

protich committed:
Merge upstream develop (rebasing DB changes)

protich committed:
Correct class name
Improve variable's preg_match_all (Thanks to Jared)

protich committed:
Remove unused functions - replaced by variable replacer

protich committed:
Add %{title} variable to internal note alert template

protich committed:
Update variables to the new format

protich committed:
Rebase DB schema signature

protich committed:
Reformat supported variables popup

protich committed:
Add ability to overwrite object variable.
 e.g %{staff.*} can be overwritten by a string (placeholder) as assign('staff', 'SYSTEM').

protich committed:
Rename patch file

protich committed:
Clobber  object on auto-cron (Fixes false credit)

protich committed:
Change postNote call

protich committed:
Add new ticket thread class. (Note, Message & Response)

protich committed:
Support new variables format
Imrove internal notes posting
Improve ticket assignment & alerts

protich committed:
Update new install SQL

protich committed:
Add transitional patch for internal (branch) installs

protich committed:
Add patch to upgrade to f4da0c9befa257b5a20a923d4e9c0e91

protich commented:
Jared - please review the PR when you get a moment.

protich committed:
Add %{comments} to variables popup window

greezybacon commented:
Peter, I don't see anything glaring. But at 1460+ line changes, it's really hard to feel totally comfortable. And what's more, this stuff is difficult and time consuming to test. We **really** need a redesign centered around test suites!

protich commented:
Most of the lines probably come from the SQL patch - I did extensive testing on the variables, a quick scan will do.
